### PR TITLE
#798 Implement profile card

### LINKED ISSUE #798: Create a way to see other group members profile
### Problem

There should be a place where group members can see others:
- Payment information;
- Bio;
- Date they joined the group;
- [Action] Send message;
- [Action] Create proposal to remove member.

### Solution

Figure out how, where and when this information is displayed:
- Add a way for users to preview their profile. Probably somewhere in the User Settings;
- [Desktop] show this information on-hover;
- [Mobile] use a drawer component;

What I have in mind is very similar to the way Slack does it. 

Next steps:
- Create mockups and get feedback!

### PR DESCRIPTION
**Pull request by pieer**

### CODE REVIEW

File: frontend/assets/style/_transitions.scss
Comment (sandrina-p) [inline frontend/assets/style/_transitions.scss]: `30px` -> `2rem`

File: frontend/views/containers/dashboard/GroupMembers.vue
Comment (sandrina-p) [inline frontend/views/containers/dashboard/GroupMembers.vue]: Tooltip is not being used in this file

File: frontend/views/components/UserName.vue
Comment (sandrina-p) [inline frontend/views/components/UserName.vue:14]: Do you plan to use this component anywhere else? As it is, it seems an extra file with low "value".

File: frontend/views/components/Tooltip.vue
Comment (sandrina-p) [inline frontend/views/components/Tooltip.vue:107]: Sweet! ✨

File: frontend/views/components/Tooltip.vue
Comment (sandrina-p) [inline frontend/views/components/Tooltip.vue]: In what scenarios would `c-tooltip` not exist? This code is executed when the tooltip is already `isActive`, so this condition would always be true

File: frontend/views/components/ProfileCard.vue
Comment (sandrina-p) [inline frontend/views/components/ProfileCard.vue]: What if we calculate `isSelf` in this same component as a property, based on `username` prop?
This component will be widely used across the app and we would DRY a lot `username === ourUsername`.

File: frontend/views/components/ProfileCard.vue
Comment (sandrina-p) [inline frontend/views/components/ProfileCard.vue]: The `paymentMethods` is already fully implemented. You can access it at `this.groupProfiles[username].paymentMethods`.

File: frontend/views/components/ProfileCard.vue
Comment (sandrina-p) [inline frontend/views/components/ProfileCard.vue]: There's no need for `removeMember` method. It can be:
```
@click='openModal("RemoveMember", { username })'
```

File: frontend/views/components/ProfileCard.vue
Comment (sandrina-p) [inline frontend/views/components/ProfileCard.vue]:
```
return !!this.groupProfiles[username]
```

File: frontend/views/components/ProfileCard.vue
Comment (sandrina-p) [inline frontend/views/components/ProfileCard.vue]: use `!!` to return a Bool as you did in `hasIncomeDetails`.

File: frontend/views/components/ProfileCard.vue
Comment (sandrina-p) [inline frontend/views/components/ProfileCard.vue]: `0 0.5rem 1.25rem`

File: frontend/views/components/ProfileCard.vue
Comment (sandrina-p) [inline frontend/views/components/ProfileCard.vue:9]: The designs show a light underline style when hovering the ProfileCard trigger (`slot`).

We have an accessibility problem here: We can't reach this component using the keyboard. I will ping @mmbotelho on Figma to double check the focus states and cc you.

On the technical side, the strategy to do an accessible "popover" menu is very similar to the one used for modals. On trigger click move the focus to the popover and then trap the keyboard navigation inside it. (See #778). I can help you if you need! Slack (browser version) do it perfectly. It's a great reference as A11Y best practices. ✨

**Review by sandrina-p — COMMENTED** [review]
Yey @pieer this seems great!! Love the smoothness :D

My only concern is around accessibility. Besides that I didn't find any other major problem.

P.S. This PR seems to close #798. Don't forget to add it to the PR description so it gets closed automatically.

File: frontend/views/components/UserName.vue
Comment (pieer) [reply frontend/views/components/UserName.vue:14]: Yes, everywhere we have the combination of name or name and displayName, which is in many places

**Review by mmbotelho — CHANGES_REQUESTED** [review]
Here's what I found @pieer :

Captura de ecrã 2020-04-14, às 11 53 39

Captura de ecrã 2020-04-14, às 11 46 04

Captura de ecrã 2020-04-14, às 11 56 43
Captura de ecrã 2020-04-14, às 11 49 53

Suggestion:
- Can we make the animation similar to the one we have when the user settings (or any fullscreen modal) appears? 

Question:
- How can I test the versions of the card that don't have the payment methods there? Are you planning on doing that in this PR or in a future one? I'm ok with either, I just need to know if I should review them now or not 😄

**Review by mmbotelho — DISMISSED** [review]
Approved @pieer ! Great job on this one!

File: frontend/views/containers/navigation/Profile.vue
Comment (taoeffect) [inline frontend/views/containers/navigation/Profile.vue]: (minor) can be written as:

```
{{ userDisplayName || ourUsername }}
```

File: frontend/views/containers/navigation/Profile.vue
Comment (sandrina-p) [inline frontend/views/containers/navigation/Profile.vue:15]: This `<span />` needs to be a `<button />`, so that someone with a keyboard can interact with it and open the profile card.

The same should happen on the members' list. A way to detect these issues is by using the keyboard only and try to open and close the profile card. Use `Tab` to go forward and `Shift + Tab` to go backward.

File: frontend/views/containers/access/PasswordModal.vue
Comment (sandrina-p) [inline frontend/views/containers/access/PasswordModal.vue]: `return false` isn't needed, or is it?

File: frontend/views/components/modal/ModalBaseTemplate.vue
Comment (sandrina-p) [inline frontend/views/components/modal/ModalBaseTemplate.vue:23]: Nice add! I completely forgot about this one.

File: frontend/views/components/UserName.vue
Comment (sandrina-p) [inline frontend/views/components/UserName.vue]: The default text size is `$size_4` (14px / 0.875rem), not 1rem.

File: frontend/views/components/Tooltip.vue
Comment (sandrina-p) [inline frontend/views/components/Tooltip.vue]: This works great on first try but if we close the profile card and then open it again it doesn't focus anymore. 🐛

File: frontend/views/components/ProfileCard.vue
Comment (sandrina-p) [inline frontend/views/components/ProfileCard.vue]: One major requirement to create an effective trapfocus system is to do it within a "dialog" element. As done in modal, we need to do it here too: `role="dialog"` [role dialog MDN docs](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/dialog_role) and it needs an accessible label. Here it can be `aria-label="{username} profile"`

**Review by sandrina-p — CHANGES_REQUESTED** [review]
Halfway there! Once we can open and close a profile card using the keyboard only, this is ready to go :D

File: frontend/views/components/Tooltip.vue
Comment (sandrina-p) [inline frontend/views/components/Tooltip.vue]: This close button is part of the profile card, not the tooltip. Why did you choose to add it to the tooltip instead?

-- EDIT 1 --
It can't be outside the profile card, otherwise, we won't be able to close the tooltip using the keyboard. I suggest to add it to the profileCard as the last element

-- EDIT 2 --
I was about to ask to also move ".c-background" to ProfileCard but then we couldn't style it correctly :/ So, for now, let's keep it as it is under the `manual` prop. In the future, if needed we improve it.

File: frontend/views/components/ProfileCard.vue
Comment (sandrina-p) [inline frontend/views/components/ProfileCard.vue]: A `<button>` does not need `.is-unstyled` class. Actually this breaks the :hover/:focus styling.

File: frontend/views/containers/navigation/Profile.vue
Comment (sandrina-p) [inline frontend/views/containers/navigation/Profile.vue]: The DS explicitly say the links should have `border-bottom` instead of `text-decoration`. The same should be made at `GroupMembers.vue`

File: frontend/views/components/Tooltip.vue
Comment (sandrina-p) [inline frontend/views/components/Tooltip.vue]: This event should be only added/removed when the tooltip opens/closes. The best place to do it is inside the directive `appendToBody`.

File: frontend/views/components/Tooltip.vue
Comment (sandrina-p) [reply frontend/views/components/Tooltip.vue]: @pieer I found the issue and fixed it!

So, you were right: `this.$slots.tooltip[0].elm` is the problem. It's correct (the actual element) on the first try but after it unmounts, it's lost. As you said, I think it's because each time the element is appended to the body, it creates a new instance, but the `$slot` elm is still the original, so it's not exactly the same element. That's why it can't focus it.

If we do this logic (keydown eventlistener and to focus) directly inside the directive `appendToBody`, it works, because there we are dealing directly with the element, instead of a Vue $slot "shallow copy".

One extra note: When the element is created, we should not `focusOnFirst`. Instead we should focus the element itself as we do in the modal using `v-focus`. 

Anyway... all the changes needed to make this work 100% for keyboard users and screen readers are done in a single commit at [this branch (a copy of yours)](https://github.com/okTurtles/group-income-simple/commit/8e1a889e5a750b1d4e96c00272a8e8f52fd45f2b). Take special attention to the HTML attributes added such as `tabindex`, `role`, and `aria-label`. They are important.

If you need help to understand something in that code, I'm free to help you in a video call :)

**Review by sandrina-p — COMMENTED** [review]
@pieer, I've found the bug cause and proposed a solution! See comments

File: frontend/utils/trapFocus.js
Comment (sandrina-p) [inline frontend/utils/trapFocus.js]: This line can be removed. We do not need to focus the first focusable element, only the el itself.

File: frontend/views/components/Tooltip.vue
Comment (sandrina-p) [inline frontend/views/components/Tooltip.vue]: The resize event should be removed always, even if there isn't a `parentNode`. Otherwise, we'll have a memory leak. And it should be `$this` instead of `this`.

File: frontend/views/components/ProfileCard.vue
Comment (taoeffect) [reply frontend/views/components/ProfileCard.vue]: Agreed makes more sense to calculate as a computed property in this component 👍

File: frontend/views/components/ProfileCard.vue
Comment (taoeffect) [reply frontend/views/components/ProfileCard.vue]: Yep, feel free to implement this part!

File: frontend/views/components/ProfileCard.vue
Comment (taoeffect) [inline frontend/views/components/ProfileCard.vue]: FYI in #900 (Payments) this is going to change, there will be a `.status` property on the groupProfile that you can check, e.g. `return this.groupProfiles[this.username].status === INVITE_STATUS.VALID`.

So either let's wait for that to be merged and update this, and/or add a `// TODO:` comment to that effect here to not forget to update this.

**Review by taoeffect — CHANGES_REQUESTED** [review]
Great job @pieer! Looks almost perfect from my testing of it. In addition to what @sandrina-p noticed, I wrote some minor comments.

One issue I did notice had to do with the animation. During testing I noticed a weird "jump" behavior, which I recorded here:

out

**Review by sandrina-p — DISMISSED** [review]
Yey! All good! The final touch with a fade-in was great!

**Review by mmbotelho — DISMISSED** [review]
My last approval was stale, but I still approve this! 😄

File: frontend/views/components/ProfileCard.vue
Comment (taoeffect) [inline frontend/views/components/ProfileCard.vue]: We should show a user's payment details regardless of whether or not we've filled in our own details.

**Review by taoeffect — CHANGES_REQUESTED** [review]
Excellent work @pieer! You fixed the animation jitter I found before. This time I found two bugs and a UX issue:

### Bug 1: Showing incorrect payment details

Screen Shot 2020-05-05 at 12 50 25 PM

In this screenshot it shows "sam" as having bitcoin information, but that's wrong, because sam didn't enter their income details at all (didn't say they're pledging or receiving).

The bitcoin information displayed belongs to use "greg".

### Bug 2: Not showing payment details when they should be shown

Screen Shot 2020-05-05 at 12 50 44 PM

Screen Shot 2020-05-05 at 12 54 40 PM

As I played around with the app, user "joe" was shown as either not having any payment details (first screenshot above, which is what you see when looked at from "joe"'s perspective, who has not entered any payment details), or having the wrong payment details as shown in the 2nd screenshot above (when seen from user "greg"'s perspective).

The correct payment details of "Paypal: super@paypal.com" aren't shown.

### UX Issue: animation for user profile popup too slow

When bringing up the profile drawer on mobile, the profile takes too long to appear. Can you please reduce the animation time to make it appear nearly instantly? It should be about 100ms or so.

Am going to continue looking through the code and posting comments. Just getting this review to you as quickly as I can so that you can start working on it sooner than later. I'll let you know either here or on Slack (or both) when I'm done reviewing all the code, which should be in an hour or so.

EDIT: also, please add two tests to catch those bugs ☝️ so that they don't return in the future and so that we ensure that the profile cards will display the correct information.

**Comment by taoeffect** [comment]
@pieer can you add tests to catch those two bugs also?

File: frontend/views/components/Tooltip.vue
Comment (taoeffect) [inline frontend/views/components/Tooltip.vue]: My understanding is that `removeEventListener` must be called on the exact saved instance of a function.

Here it's being called on a new instance of a function, `e => $this.trapFocus(e, el)`, instead of a reference to an existing instance. This will result in a memory leak so it must be fixed.

File: frontend/views/containers/access/PasswordModal.vue
Comment (taoeffect) [inline frontend/views/containers/access/PasswordModal.vue]: This clearly doesn't belong here. Not sure why this file was modified.

File: frontend/views/containers/dashboard/GroupMembers.vue
Comment (taoeffect) [inline frontend/views/containers/dashboard/GroupMembers.vue]: I think @sandrina-p mentioned (and I agreed) that this `isSelf` property should be removed. Instead the profile card should have a computed property to do this.

**Comment by taoeffect** [comment]
OK, finished code review as well, please see new comments.

Also, I noticed this wasn't implemented, yet @mmbotelho gave approval...

Screen Shot 2020-05-05 at 1 04 07 PM

**Review by mmbotelho — COMMENTED** [review]
This is essentially the same thing @taoeffect pointed out, but I just want to add these screenshots with implemented and designed side by side,  so it's easier to understand:

Captura de ecrã 2020-05-06, às 11 41 26
Captura de ecrã 2020-05-06, às 11 40 18

**Review by sandrina-p — CHANGES_REQUESTED** [review]
Can't wait to see this on master 🚀

**Review by sandrina-p — DISMISSED** [review]

File: frontend/views/containers/access/PasswordModal.vue
Comment (taoeffect) [inline frontend/views/containers/access/PasswordModal.vue]: Unused code should be removed.

File: frontend/views/components/ProfileCard.vue
Comment (taoeffect) [inline frontend/views/components/ProfileCard.vue]: `name` is the index, so this should be `paymentMethod.name`

**Review by taoeffect — CHANGES_REQUESTED** [review]
Excellent work @pieer!

Just a few issues remaining:

- [This comment from the previous review about `removeEventListener`](https://github.com/okTurtles/group-income-simple/pull/884/files#r420399119) still needs to be addressed
- The speed of the mobile drawer appearing still needs to be made faster (I think 2x faster would keep the UI feeling snappy and not sluggish)
- There's a bug where a number is shown instead of the payment method:

Screen Shot 2020-05-08 at 2 09 29 PM

File: frontend/views/components/ProfileCard.vue
Comment (sandrina-p) [reply frontend/views/components/ProfileCard.vue]: Good catch. Probably @pieer didn't get the change to see this PR #912 that changed how the payments are stored, what's why this broke the page.

**Review by mmbotelho — APPROVED** [review]
Perfect @pieer !

**Review by taoeffect — APPROVED** [review]
Fantastic work @pieer, the new Profile card looks and works great! 🥇 💯

**Pull request merged**
